### PR TITLE
feat(compose/agent): add stack.agent.ports option in config.yml to publish ports for agent

### PR DESCRIPTION
**Pull request by pkoutsovasilis**
This PR introduces support for defining inside `config.yml` of the profile ports for agent service to be published

- Relates to https://github.com/elastic/elastic-package/issues/1262
- Relates to https://github.com/elastic/elastic-package/issues/1355

### CODE REVIEW

**Comment by jsoriano** [comment]
I wonder if https://github.com/elastic/elastic-package/pull/1795 supersedes this one.

@pkoutsovasilis is there any use case that would not be covered by defining the ports in the system test cases?

**Comment by pkoutsovasilis** [comment]
> I wonder if https://github.com/elastic/elastic-package/pull/1795 supersedes this one.
> 
> @pkoutsovasilis is there any use case that would not be covered by defining the ports in the system test cases?

hmmm interesting 🤔 the answer is yes when you are developing a new integration or testing something about an existing one. Sure this should end up in a system test but before you have a test you have to do certain manual steps to get there. Also, since this PR is now open for 6 days, if you need to extend this one to custom agents deployments as this https://github.com/elastic/elastic-package/pull/1795 (2 days ago) does, wouldn't make sense to ask for it here than open a complete different PR?

**Comment by jsoriano** [comment]
> the answer is yes when you are developing a new integration or testing something about an existing one.

Makes sense. This is probably also useful in cases where the stack is used for manual testing.

> since this PR is now open for 6 days, if you needed support for custom agents deployments as this #1795 (2 days ago) does, wouldn't make sense to ask for it here than open a complete different PR?

Well, they likely come from two different needings, we probably need both approaches. The one here for manual tests, and the other one for testing and CI in the new custom agents.

**Comment by pkoutsovasilis** [comment]
> Makes sense. This is probably also useful in cases where the stack is used for manual testing.
> 
> >

and for manual testing/development sometimes we need also custom volume mounts 😄 e.g. for this [one](https://github.com/elastic/integrations/blob/main/packages/fim/manifest.yml) we need this `/sys/kernel/debug:/sys/kernel/debug`

File: internal/stack/_static/docker-compose-stack.yml.tmpl [hunk 1]
```diff
@@ -139,6 +139,9 @@ services:
       interval: 5s
     hostname: docker-fleet-agent
     env_file: "./elastic-agent.env"
+    {{- if eq (fact "agent_has_publish_ports") "true" }}
+    ports: {{ fact "agent_publish_ports" }}
+    {{- end }}
```
Comment (jsoriano) [inline internal/stack/_static/docker-compose-stack.yml.tmpl:144]: Nit. Maybe we can pass here the comma separated list of ports, without the brackets, and check here if the list is empty? This way we would not need the `has_publish_ports` fact.
```suggestion
    {{- if ne (fact "agent_publish_ports") "" }}
    ports: {{ fact "agent_publish_ports" }}
    {{- end }}
```

File: internal/stack/resources.go [hunk 2]
```diff
@@ -152,9 +152,25 @@ func applyResources(profile *profile.Profile, stackVersion string) error {
...
+		facter["agent_publish_ports"] = fmt.Sprintf("[%s]", strings.Join(agentPorts, ","))
+	} else {
+		facter["agent_has_publish_ports"] = "false"
+	}
+
+	resourceManager.AddFacter(facter)
+
+	if err := os.MkdirAll(stackDir, 0755); err != nil {
+		return fmt.Errorf("failed to create stack directory: %w", err)
+	}
```
Comment (jsoriano) [inline internal/stack/resources.go:173]: This should not be needed :thinking:

File: internal/profile/profile.go [hunk 3]
```diff
@@ -141,13 +141,13 @@ type Profile struct {
 }
 
 // Path returns an absolute path to the given file
-func (profile Profile) Path(names ...string) string {
+func (profile *Profile) Path(names ...string) string {
```
[the remark this answers is not in the document]
Comment (pkoutsovasilis) [reply internal/profile/profile.go:144]: so here it depends, it is considered an unusual pattern to have a struct definition that mixes both value and pointer receivers, and from a quick look at the code, profile is widely used as a pointer. Thus I homogenise the struct by making these pointer receivers but if there is a reason behind this divergence happy revert this change

File: internal/stack/resources.go (hunk 2 above)
Comment (pkoutsovasilis) [reply internal/stack/resources.go:165]: I just added the err, `os.MkdirAll...` was already there but without being checked. If we want the error to be ignored there then I would prefer have it as `_ = os.MkdirAll...` so it is evident that we explicitly ignore this 🙂

**Comment by pkoutsovasilis** [comment]
> Well, they likely come from two different needings, we probably need both approaches. The one here for manual tests, and the other one for testing and CI in the new custom agents.

the more I look the agentdeployer the more it makes sense to me expose this through cli and someone be able to deploy "custom" agents in the stack and focus any extendability effort there, @jsoriano what do you think?

File: internal/profile/config.go [hunk 4]
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/elastic/go-ucfg/yaml"
 
 	"github.com/elastic/elastic-package/internal/common"
+
+	"github.com/mitchellh/mapstructure"
```
Comment (jsoriano) [inline internal/profile/config.go:16]: Nit. Imports order.
```suggestion
	"github.com/elastic/go-ucfg/yaml"
	"github.com/mitchellh/mapstructure"

	"github.com/elastic/elastic-package/internal/common"
```

File: internal/profile/profile.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply internal/profile/profile.go:144]: Good point. I think I would rather use the profile by value, but it is true that it is mostly used as pointer now, so as you prefer.

File: internal/profile/config.go [hunk 5]
```diff
@@ -48,3 +50,18 @@ func (c *config) get(name string) (string, bool) {
 		return fmt.Sprintf("%v", v), true
 	}
 }
+
+func (c *config) Decode(name string, out any) error {
```
Comment (jsoriano) [inline internal/profile/config.go:54]: Please consider adding a test for this.

File: internal/stack/resources.go (hunk 2 above)
Comment (jsoriano) [reply internal/stack/resources.go:165]: It is fine.

**Review by jsoriano — APPROVED** [review]
Thanks! I think that with this change and with https://github.com/elastic/elastic-package/pull/1795, we can close https://github.com/elastic/elastic-package/issues/1355.

**Comment by jsoriano** [comment]
> the more I look the agentdeployer the more it makes sense to me expose this through cli and someone be able to deploy "custom" agents in the stack and focus any extendability effort there, @jsoriano what do you think?

Yeah, I see the agent deployer as the future to deploy any kind of agents. There is still some work to do in this area though, @mrodm is now working on this.

In any case we have to keep in mind that the purpose of `elastic-package` is to be the main tool for package developers. We should be cautious about adding features that go beyond package development.

**Review by jsoriano — APPROVED** [review]
:+1:

**Pull request merged**